File: crates/cli/src/data/brands.rs
use crate::data::common::{Contacts, Socials};
use jsonschema::{Draft, JSONSchema};
use schemars::schema::RootSchema;
use schemars::{schema_for, JsonSchema};
use serde_json::Value;
use std::str;

#[derive(Debug, Deserialize, Clone, JsonSchema)]
pub struct Brand {
    pub name: String,
    #[serde(rename = "registeredCompanyName")]
    pub registered_company_name: String,
    #[serde(rename = "groupName")]
    pub group_name: Option<String>,
    pub description: Option<String>,
    pub address: Option<Address>,
    #[serde(rename = "contactInfo")]
    pub contact_info: Option<Contacts>,
    pub social: Option<Socials>,
    pub kind: Kind,
    pub status: Status,
    pub version: u8,
}

#[derive(Debug, Deserialize, Clone, JsonSchema)]
pub enum Kind {
    #[serde(rename = "INDUSTRIAL")]
    Industrial,
    #[serde(rename = "BRASS_MODELS")]
    BrassModels,
}

#[derive(Debug, Deserialize, Clone, JsonSchema)]
pub enum Status {
    #[serde(rename = "ACTIVE")]
    Active,
    #[serde(rename = "OUT_OF_BUSINESS")]
    OutOfBusiness,
}

#[derive(Debug, Deserialize, Clone, JsonSchema)]
pub struct Address {
    #[serde(rename = "streetAddress")]
    pub street_address: String,
    #[serde(rename = "extendedAddress")]
    pub extended_address: Option<String>,
    #[serde(rename = "postalCode")]
    pub postal_code: String,
    pub city: String,
    pub region: Option<String>,
    pub country: String,
}

impl str::FromStr for Brand {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let schema: RootSchema = schema_for!(Brand);
        let schema: Value = serde_json::from_str(&serde_json::to_string_pretty(&schema).unwrap()).unwrap();

        let compiled = JSONSchema::options()
            .with_draft(Draft::Draft7)
            .compile(&schema)
            .expect("A valid schema");
        let brand: Value = serde_json::from_str(&s).unwrap();

        let result = compiled.validate(&brand);
        match result {
            Ok(()) => {
                let brand: Brand = serde_json::from_str(&s).unwrap();
                Ok(brand)
            }
            Err(errors) => {
                for error in errors {
                    println!("{:?}", error);
                }
                Err("".to_owned())
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    mod brands {
        use super::*;
        use std::str::FromStr;

        #[test]
        fn it_should_parse_yaml_files() {
            let contents = r#"
{
  "name": "ACME",
  "registeredCompanyName": "Associazione Costruzioni Modellistiche Esatte",
  "groupName": null,
  "description": null,
  "address": {
    "streetAddress": "Viale Lombardia, 27",
    "extendedAddress": null,
    "postalCode": "20131",
    "city": "Milano",
    "region": "MI",
    "country": "it"
  },
  "contactInfo": {
    "email": "mail@acmetreni.com",
    "websiteUrl": "http://www.acmetreni.com",
    "phone": null
  },
  "social": {
    "instragram": null,
    "facebook": null,
    "youtube": null,
    "twitter": null
  },
  "kind": "INDUSTRIAL",
  "status": "ACTIVE",
  "version": 1
}
            "#;

            let result = Brand::from_str(contents);
            assert!(result.is_ok());
        }
    }
}

File: e2e/api/src/main/kotlin/io/github/carlomicieli/trenako/database/seeding/Brands.kt
/*
 *   Copyright (c) 2022-2023 (C) Carlo Micieli
 *
 *    Licensed to the Apache Software Foundation (ASF) under one
 *    or more contributor license agreements.  See the NOTICE file
 *    distributed with this work for additional information
 *    regarding copyright ownership.  The ASF licenses this file
 *    to you under the Apache License, Version 2.0 (the
 *    "License"); you may not use this file except in compliance
 *    with the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing,
 *    software distributed under the License is distributed on an
 *    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 *    KIND, either express or implied.  See the License for the
 *    specific language governing permissions and limitations
 *    under the License.
 */
package io.github.carlomicieli.trenako.database.seeding

import io.github.carlomicieli.trenako.database.BrandRow
import io.github.carlomicieli.trenako.database.BrandsRepository
import io.github.carlomicieli.trenako.model.Address
import io.github.carlomicieli.trenako.model.BrandKind
import io.github.carlomicieli.trenako.model.BrandStatus
import io.github.carlomicieli.trenako.model.ContactInfo
import io.github.carlomicieli.trenako.model.OrganizationEntityType
import io.github.carlomicieli.trenako.model.Socials
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import java.net.URI
import java.time.Instant

@Component
class Brands(val brandsRepository: BrandsRepository) {
    companion object {
        val LOG: Logger = LoggerFactory.getLogger(Brands::class.java)

        val ACME: BrandRow = BrandRow(
            brandId = "acme",
            name = "ACME",
            registeredCompanyName = "Associazione Costruzioni Modellistiche Esatte",
            organizationEntityType = OrganizationEntityType.LIMITED_COMPANY,
            groupName = "UNKNOWN",
            description = mapOf("it" to "descrizione", "en" to "description"),
            kind = BrandKind.INDUSTRIAL,
            status = BrandStatus.ACTIVE,
            contacts = ContactInfo()
                .email("mail@mail.com")
                .phone("+14152370800")
                .websiteUrl(URI.create("https://www.acmetreni.com")),
            address = Address()
                .streetAddress("Viale Lombardia, 27")
                .city("Milano")
                .country("IT")
                .region("MI")
                .postalCode("20131"),
            socials = Socials()
                .facebook(URI.create("facebook_handler"))
                .instagram(URI.create("instagram_handler"))
                .linkedin(URI.create("linkedin_handler"))
                .twitter(URI.create("twitter_handler"))
                .youtube(URI.create("youtube_handler")),
            version = 0,
            createdAt = Instant.parse("2023-04-21T10:15:30.00Z"),
            lastModifiedAt = Instant.parse("2023-04-21T10:15:30.00Z")
        )
    }

    suspend fun seed() {
        LOG.info("Seeding the brands")

        brandsRepository.insert(ACME)
        LOG.info("Inserted 1 brand...")
    }
}
